### DISCUSSION TITLE
How to use pydantic correctly in Mapped dataclasses?

### DISCUSSION THREAD

**Discussion by dunossauro** [sqlalchemy/sqlalchemy#9717] [Usage Questions]
I'm looking to the simple example using [`MappedAsDataclass` and pydantic](https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#integrating-with-alternate-dataclass-providers-such-as-pydantic) and trying to increment a little

```python
from pydantic.dataclasses import dataclass
from typing import Optional

class Base(
    MappedAsDataclass,
    DeclarativeBase,
    dataclass_callable=dataclass,
):
    ...


class Todo(Base):
    __tablename__ = 'todos'

    id: Mapped[int] = mapped_column(init=False, primary_key=True, autoincrement=True)
    task: Mapped[str]
    state: Mapped[str] = mapped_column(default='todo')
```

but when I call `Todo(task=1, state=1)` receive an error:

```python
ValidationError: 1 validation error for Todo
id
  field required (type=value_error.missing)
````

So, I try use id for test:

```python
>>> Todo(id=2, task=1, state=1)

TypeError: Todo.__init__() got an unexpected keyword argument 'id'
```

--- 

I change the id definition. Now works work when explicit call id

```python
from typing import Optional
...
id: Mapped[Optional[int]] = mapped_column(primary_key=True, autoincrement=True)

>>> Todo(id=2, task='1', state='1')
Todo(id=2, task='1', state='1')

>>> Todo(task=1, state=1)
TypeError: Todo.__init__() missing 1 required positional argument: 'id'
```

But this is not a correct behavior. How to define correctly this column?

For debug purposes, i'm using this version of packages:
sqlalchemy = "^2.0.11"
pydantic = "^1.10.7"

**Comment by zzzeek** [comment]
hi - 

FTR this mapping for me doesn't even succeed if I include `from __future__ import annotations` at the top; pydantic seems to be applying unexpected transformations to the annotations that cause our mapping to break (turning types into forward refs).    When this directive is removed, I'm not sure what goes wrong, ~~but it's likely something similar~~ nope nothing wrong.

The presence of the "alternate provider" documented at https://docs.sqlalchemy.org/en/20/orm/dataclasses.html#integrating-with-alternate-dataclass-providers-such-as-pydantic was at the request of a user that assured me pydantic's dataclass function works identically to the standard library one ~~however this seems to not be the case~~ it works, it just needs more formalities in the declaration.       ~~Unfortunately though I can't support pydantic itself so if pydantic is doing things that make this library feature infeasible, I will likely update the docs to indicate that this hook is not known to be working with pydantic at this time.~~

~~im going to probably update the docs now anyway as this really seems to not be working~~

**Reply by zzzeek** [reply]
I get the same error without using SQLAlchemy.  ~~It seems like pydantic does not recognize dataclasses.field() ?~~ (it does)   I'm not sure what is supposed to be used instead (thought this was a drop-in)

```py
from pydantic.dataclasses import dataclass as pydantic_dataclass
import dataclasses


@dataclasses.dataclass
class TodoDC:
    id: int = dataclasses.field(init=False)
    task: str
    state: str

TodoDC(task="1", state="1")


@pydantic_dataclass
class TodoPD:

    id: int = dataclasses.field(init=False)
    task: str = dataclasses.field()
    state: str = dataclasses.field()

TodoPD(task="1", state="1")
```

same error fro the second one

**Reply by zzzeek** [reply]
I have a patch going through that will fix the fwd ref issue, but as far as "id" above, I dont have  a fix there other than passing an adaption function that sends pydantic what it recognizes.

**Reply by CaselIT** [reply]
> I will likely update the docs to indicate that this hook is not known to be working with pydantic at this time.

It kinda works, but you need to respect the somewhat wired rules pydantic has

**Comment by CaselIT** [comment]
Hi,

I think you need to use this to make pydantic happy
```py
id: Mapped[Optional[int]] = mapped_column(primary_key=True, autoincrement=True, init=False)
```

In your first example it's complaining since id is declared as `int` but it was not assigned so the error.

**Reply by zzzeek** [reply]
wow.    OK.    Then I guess with my one fix I have in review we still can support pydantic.  ill revert all that doomsaying language

**Reply by zzzeek** [reply]
uh nope, not working

**Reply by zzzeek** [reply]
does not work:

```py
from pydantic.dataclasses import dataclass as pydantic_dataclass
from typing import Optional
import dataclasses


@dataclasses.dataclass
class TodoDC:
    id: int = dataclasses.field(init=False)
    task: str
    state: str

TodoDC(task="1", state="1")


@pydantic_dataclass
class TodoPD:

    id: Optional[int] = dataclasses.field(init=False)
    task: str = dataclasses.field()
    state: str = dataclasses.field()

TodoPD(task="1", state="1")
```

**Reply by zzzeek** [reply]
OK it has to also have "default=None":

```py
class Todo(Base):
    __tablename__ = 'todos'

    id: Mapped[Optional[int]] = mapped_column(primary_key=True, autoincrement=True, init=False, default=None)
    task: Mapped[str] = mapped_column(default='todo')
    state: Mapped[str] = mapped_column(default='todo')

```

regular dataclasses ignores that but turns out `__repr__()` fails if you dont give it a default

**Reply by CaselIT** [reply]
Yes pydantic is validating `id is int`. If not provided or set to None it will not work.
If you instead use `id: Mapped[int|None]` pydantic should be fine with None there.

I'm not sure if it actually places None there. If it does so then that's not good since I guess sqlalchemy would try to insert that None (Not sure).

I guess all this is for users of pydantic to figure out
